LuaRules/Gadgets/unit_targetting_metal_limit.lua

```diff
@@ -0,0 +1,202 @@
+		Spring.Echo("unitMetalMin" .. tostring(unitMetalMin[unitID]))
+		Spring.Echo("unitDefCost" .. tostring(unitDefCost[unitDefIDAsKnown]))
+		if unitMetalMin[unitID] and unitMetalMin[unitID] > unitDefCost[unitDefID] then
+			return true
+		end
+	end
+	return false
+end
+
+--------------------------------------------------------------------------------
+-- Command Handling 
+local function PreventFiringAtChaffToggleCommand(unitID, unitDefID, cmdParams, cmdOptions)
+	if canHandleUnit[unitID] then
+		local state = cmdParams[1] or 1
+		local cmdDescID = spFindUnitCmdDesc(unitID, CMD_MIN_METAL_TO_TARGET)
+
+		local newParams = {}
+		if (cmdDescID) then
+			for k,v in pairs(metalThreshholdsByUnitDefIDs[unitDefID]) do newParams[k] = v end
+			newParams[1] = state
+			preventChaffShootingCmdDesc.params = newParams
+			spEditUnitCmdDesc(unitID, cmdDescID, {params = preventChaffShootingCmdDesc.params})
+		end
+		Spring.SendLuaRulesMsg(minMetalMsg .. " ".. tostring(unitID) .. " " .. tostring(metalThreshholdsByUnitDefIDs[unitDefID][state + 2]))
+		return false
+	end
+	return true
+end
+
+function gadget:AllowCommand_GetWantedCommand()
+	return {[CMD_MIN_METAL_TO_TARGET] = true}
+end
+
+function gadget:AllowCommand_GetWantedUnitDefID()
+	return true
+end
+
+function gadget:RecvLuaMsg (msg, playerID)
+	if msg:sub(1, #minMetalMsg) ~= minMetalMsg then
+		return
+	end
+
+	-- Break up the message to find unidID and the new metal minimum for targetting
+	local strDividerPos = string.find(msg, " ", #minMetalMsg + 2)
+	if strDividerPos == nil then return end
+	local unitID = tonumber(msg:sub(#minMetalMsg + 2, strDividerPos))
+	if unitID == nil then return end
+	local newMetalMin = tonumber(msg:sub(strDividerPos + 1, #msg))
+	if newMetalMin == nil then return end
+
+	-- Verify this player is authorized
+	local unitTeamId = Spring.GetUnitTeam(unitID)
+	if unitTeamId == nil then return end
+	local _,_,_,playerTeamId = Spring.GetPlayerInfo(playerID)
+	if playerTeamId == nil then return end
+
+	if unitTeamId == playerTeamId then
+		unitMetalMin[unitID] = newMetalMin
+	end
+end
+
+function gadget:AllowCommand(unitID, unitDefID, teamID, cmdID, cmdParams, cmdOptions)
+	if (cmdID == CMD_MIN_METAL_TO_TARGET) then
+		return PreventFiringAtChaffToggleCommand(unitID, unitDefID, cmdParams, cmdOptions)
+	else
+		return true  -- command was not used
+	end
+end
+
+--------------------------------------------------------------------------------
+-- Unit Handling
+
+function gadget:UnitCreated(unitID, unitDefID, teamID)
+	if metalThreshholdsByUnitDefIDs[unitDefID] then
+		spInsertUnitCmdDesc(unitID, preventChaffShootingCmdDesc)
+		canHandleUnit[unitID] = true
+		PreventFiringAtChaffToggleCommand(unitID, unitDefID, metalThreshholdsByUnitDefIDs[unitDefID])
+	end
+end
+
+function gadget:UnitDestroyed(unitID)
+	if canHandleUnit[unitID] then
+		if unitMetalMin[unitID] then
+			unitMetalMin[unitID] = nil
+		end
+		canHandleUnit[unitID] = nil
+	end
```

Comment:
Looking the value up takes just as much time as setting it
```suggestion
	unitMetalMin[unitID] = nil
	canHandleUnit[unitID] = nil
```

Reply:
Doesn't that bloat canHandleUnit unnecessarily though? Taking up a (small) amount of extra memory on units that don't use metal threshhold targetting?

Reply:
As far as I know `nil` takes up no space.